packages/web-frontend/src/sagas.js

```diff
@@ -117,9 +130,124 @@ import {
   selectOrgUnit,
   selectOrgUnitChildren,
   selectOrgUnitCountry,
+  selectProjectByCode,
 } from './selectors';
 import { formatDateForApi, isMobile, processMeasureInfo } from './utils';
 import { getDefaultDates } from './utils/periodGranularities';
+import { fetchProjectData } from './projects/sagas';
+import { clearLocation } from './historyNavigation/historyNavigation';
+import { decodeLocation } from './historyNavigation/utils';
+import { PASSWORD_RESET_PREFIX, VERIFY_EMAIL_PREFIX } from './historyNavigation/constants';
+
+function* watchFetchInitialData() {
+  yield take(FETCH_INITIAL_DATA);
+
+  // Login must happen first so that projects return the correct access flags
+  yield call(findUserLoggedIn, LOGIN_TYPES.AUTO);
+  yield call(fetchProjectData);
+  yield call(handleLocationChange, {
+    location: getInitialLocation(),
+    previousLocation: clearLocation(),
+  });
+}
+
+function* handleInvalidPermission({ projectCode }) {
+  const state = yield select();
+  const { isUserLoggedIn } = state.authentication;
+
+  if (isUserLoggedIn) {
+    // show project access dialog
+    const project = selectProjectByCode(state, projectCode);
+
+    if ('code' in project) {
```

Comment:
Probably this is a bit closer to the actual check (that a project exists?) Since we don't need `project.code` in specific

```suggestion
    if (project) {
```

Reply:
Good point. I wanted to make sure `project = {}` didn't return a false positive. So I've update the check to be: `Object.keys(project).length > 0`